File: src/dirtoo/gui/leap_widget.py
# dirtoo - File and directory manipulation tools for Python
# Copyright (C) 2018 Ingo Ruhnke <grumbel@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.


from typing import TYPE_CHECKING, cast, Optional

from PyQt6.QtCore import Qt, pyqtSignal
from PyQt6.QtWidgets import (QLineEdit, QVBoxLayout, QWidget)
from PyQt6.QtGui import QFocusEvent, QKeyEvent, QShowEvent, QHideEvent

if TYPE_CHECKING:
    from dirtoo.fileview.file_view import FileView


class LeapWidget(QWidget):

    sig_leap = pyqtSignal(str, bool, bool)

    def __init__(self, parent: Optional['FileView']) -> None:
        super().__init__(parent,
                         Qt.WindowType.Window |
                         Qt.WindowType.WindowStaysOnTopHint |
                         # Qt.WindowType.X11BypassWindowManagerHint |
                         Qt.WindowType.FramelessWindowHint)

        self._build_gui()

    def _build_gui(self) -> None:
        self._line_edit = QLineEdit(self)
        vbox = QVBoxLayout()
        vbox.addWidget(self._line_edit)
        vbox.setContentsMargins(0, 0, 0, 0)
        self.setLayout(vbox)
        self.resize(200, 40)

        self._line_edit.editingFinished.connect(self.hide)
        self._line_edit.textChanged.connect(self.on_text_changed)

    def showEvent(self, ev: QShowEvent) -> None:
        super().showEvent(ev)
        self.place_widget()

    def place_widget(self) -> None:
        parent = cast('FileView', self.parentWidget())
        pos = parent.mapToGlobal(parent.viewport().geometry().bottomRight())

        self.move(pos.x() - self.width(),
                  pos.y() - self.height())

    def focusOutEvent(self, ev: QFocusEvent) -> None:
        print("focus out")
        super().focusOutEvent(ev)
        self.hide()

    def hideEvent(self, ev: QHideEvent) -> None:
        super().hideEvent(ev)

    def keyPressEvent(self, ev: QKeyEvent) -> None:
        if ev.key() in (Qt.Key.Key_Return, Qt.Key.Key_Enter):
            self.parentWidget().keyPressEvent(ev)
        else:
            super().keyPressEvent(ev)
            if ev.key() == Qt.Key.Key_Escape:
                self.hide()
            elif ev.key() == Qt.Key.Key_Up:
                self.sig_leap.emit(self._line_edit.text(), False, True)
            elif ev.key() == Qt.Key.Key_Down:
                self.sig_leap.emit(self._line_edit.text(), True, True)

    def on_text_changed(self, text: str) -> None:
        self.sig_leap.emit(text, True, False)


# EOF #

File: dirtoo/apps/dirtoo/leap_widget.cpp
// SPDX-FileCopyrightText: 2026 Ingo Ruhnke <grumbel@gmail.com>
// SPDX-License-Identifier: GPL-3.0-or-later

#include "leap_widget.hpp"

#include <QEvent>
#include <QKeyEvent>
#include <QLineEdit>
#include <QShowEvent>
#include <QVBoxLayout>

namespace dirtoo::app {

LeapWidget::LeapWidget(QWidget* parent)
    : QWidget(parent, Qt::Tool | Qt::FramelessWindowHint | Qt::WindowStaysOnTopHint)
{
  edit_ = new QLineEdit(this);
  edit_->setPlaceholderText(QStringLiteral("Jump to…"));
  auto* layout = new QVBoxLayout(this);
  layout->setContentsMargins(4, 4, 4, 4);
  layout->addWidget(edit_);
  resize(220, 40);

  connect(edit_, &QLineEdit::textChanged, this, [this](const QString& text) {
    emit leap(text, true, false);
  });
  connect(edit_, &QLineEdit::returnPressed, this, [this] {
    emit leap(edit_->text(), true, true);
    hide();
  });
  edit_->installEventFilter(this);
}

void LeapWidget::clear()
{
  edit_->clear();
}

QString LeapWidget::text() const
{
  return edit_->text();
}

void LeapWidget::show_and_focus()
{
  clear();
  show();
  place_on_parent();
  edit_->setFocus(Qt::PopupFocusReason);
}

void LeapWidget::place_on_parent()
{
  QWidget* p = parentWidget();
  if (p == nullptr) {
    return;
  }
  const QPoint bottom_right = p->mapToGlobal(p->rect().bottomRight());
  move(bottom_right.x() - width() - 8, bottom_right.y() - height() - 8);
}

void LeapWidget::showEvent(QShowEvent* event)
{
  QWidget::showEvent(event);
  place_on_parent();
}

void LeapWidget::keyPressEvent(QKeyEvent* event)
{
  if (event->key() == Qt::Key_Escape) {
    hide();
    event->accept();
    return;
  }
  QWidget::keyPressEvent(event);
}

bool LeapWidget::eventFilter(QObject* obj, QEvent* event)
{
  if (obj == edit_ && event->type() == QEvent::KeyPress) {
    auto* ke = static_cast<QKeyEvent*>(event);
    if (ke->key() == Qt::Key_Escape) {
      hide();
      return true;
    }
    if (ke->key() == Qt::Key_Up) {
      emit leap(edit_->text(), false, true);
      return true;
    }
    if (ke->key() == Qt::Key_Down) {
      emit leap(edit_->text(), true, true);
      return true;
    }
  }
  if (obj == edit_ && event->type() == QEvent::FocusOut) {
    // Delay hide so clicks on the list still work; simple: hide on focus out.
    hide();
  }
  return QWidget::eventFilter(obj, event);
}

} // namespace dirtoo::app
